drivers/led_strip/ws2812_gpio.c

```diff
@@ -0,0 +1,246 @@
+ * 2. GPIO set/clear
+ */
+
+/*
+ * T1H: 1 bit high pulse delay: 12 cycles == .75 usec
+ * T0H: 0 bit high pulse delay: 4 cycles == .25 usec
+ * TxL: inter-bit low pulse delay: 8 cycles == .5 usec
+ *
+ * We can't use k_busy_wait() here: its argument is in microseconds,
+ * and we need roughly .05 microsecond resolution.
+ */
+#define DELAY_T1H "nop\nnop\nnop\nnop\nnop\nnop\nnop\nnop\nnop\nnop\nnop\nnop\n"
+#define DELAY_T0H "nop\nnop\nnop\nnop\n"
+#define DELAY_TxL "nop\nnop\nnop\nnop\nnop\nnop\nnop\nnop\n"
+
+/*
+ * GPIO set/clear (these make assumptions about assembly details
+ * below).
+ *
+ * This uses OUTCLR == OUTSET+4.
+ *
+ * We should be able to make this portable using the results of
+ * https://github.com/zephyrproject-rtos/zephyr/issues/11917.
+ *
+ * We already have the GPIO device stashed in ws2812_gpio_data, so
+ * this driver can be used as a test case for the optimized API.
+ *
+ * Per Arm docs, both Rd and Rn must be r0-r7, so we use the "l"
+ * constraint in the below assembly.
+ */
+#define SET_HIGH "str %[p], [%[r], #0]\n" /* OUTSET = BIT(LED_PIN) */
+#define SET_LOW "str %[p], [%[r], #4]\n"  /* OUTCLR = BIT(LED_PIN) */
+
+/* Send out a 1 bit's pulse */
+#define ONE_BIT(base, pin) do {			\
+	__asm volatile (SET_HIGH			\
+			DELAY_T1H			\
+			SET_LOW			\
+			DELAY_TxL			\
+			::				\
+			[r] "l" (base),		\
+			[p] "l" (pin)); } while (0)
+
+/* Send out a 0 bit's pulse */
+#define ZERO_BIT(base, pin) do {			\
+	__asm volatile (SET_HIGH			\
+			DELAY_T0H			\
+			SET_LOW			\
+			DELAY_TxL			\
+			::				\
+			[r] "l" (base),		\
+			[p] "l" (pin)); } while (0)
+
+static int send_buf(struct device *dev, u8_t *buf, size_t len)
+{
+	volatile u32_t *base = (u32_t *)&NRF_GPIO->OUTSET;
+	const u32_t val = BIT(dev_cfg(dev)->pin);
+	struct device *clk = dev_data(dev)->clk;
+	unsigned int key;
+	int rc;
+
+	rc = clock_control_on(clk, NULL);
+	if (rc) {
+		return rc;
+	}
+
+	key = irq_lock();
+
+	while (len--) {
+		u32_t b = *buf++;
+		s32_t i;
+
+		/*
+		 * Generate signal out of the bits, MSbit first.
+		 *
+		 * Accumulator maintenance and branching mean the
+		 * inter-bit time will be longer than TxL, but the
+		 * wp.josh.com blog post says we have at least 5 usec
+		 * of slack time between bits before we risk the
+		 * signal getting latched, so this will be fine as
+		 * long as the compiler does something minimally
+		 * reasonable.
+		 */
+		for (i = 7; i >= 0; i--) {
+			if (b & BIT(i)) {
+				ONE_BIT(base, val);
+			} else {
+				ZERO_BIT(base, val);
+			}
+		}
+	}
+
+	irq_unlock(key);
+
+	rc = clock_control_off(clk, NULL);
+
+	return rc;
+}
+
+static int ws2812_gpio_update_rgb(struct device *dev, struct led_rgb *pixels,
+				  size_t num_pixels)
+{
+	const struct ws2812_gpio_cfg *config = dev->config->config_info;
+	const bool has_white = config->has_white;
+	u8_t *ptr = (u8_t *)pixels;
+	size_t i;
+
+	/* Convert from RGB to on-wire format (GRB or GRBW) */
+	for (i = 0; i < num_pixels; i++) {
```

Comment:
`for (size_t i = 0; i < num_pixels; i++) {` ?

Reply:
I prefer to keep this in the traditional Linux kernel style.